### PR TITLE
Extend test matrix to cover more possible build combinations of features

### PR DESCRIPTION
**Pull request by Jakuje**
With the amount of features the kryoptic supports, keeping track that they are self contained and all of the dependencies is quite tedious work.

This extends the test matrix and updates the tests to be able to run with different storages. For now, `jsondb`, `sqlitedb` and `nssdb` work, but the `memorydb` does not. Additionally, I encountered the issue that the memory db is not possible to provision, unless encryption is turned on, because it tries to find PIN counters, which are not in non-encrypted storage.

Fixes: #118

### CODE REVIEW

**Comment by Jakuje** [comment]
This is mostly ready for review. The only thing to figure out is the issue with the memorydb and why the jsondb fails in CI, but works locally.
If needed, changes can be merged and CI changes could land later after we will figure out last details.

File: src/tests/aes.rs [hunk 1]
```diff
@@ -500,14 +514,10 @@ fn test_aes_operations() {
...
-            Err(ret) => return assert_eq!(ret, CKR_OK),
-        };
-        let plaintext = match get_test_data(session, testname, "plaintext") {
-            Ok(vec) => vec,
-            Err(ret) => return assert_eq!(ret, CKR_OK),
-        };
+        let ciphertext = hex::decode("4154c0be71072945d8156f5f046d198d")
+            .expect("Failed to decode ciphertext");
+        let plaintext = hex::decode("8b2b1b22f733ac09d1196d6be6a87a72")
+            .expect("Failed to decode plaintext");
```
Comment (simo5) [inline src/tests/aes.rs:520]: What about turning get_test_data to be a function like you did for get_gcm_test_data() ?
It can be a function that return data from an array of test data base on an index or something.
Also in the commit message CKA_DATA -> CKO_DATA

File: src/tests/init.rs [hunk 2]
```diff
@@ -41,7 +41,12 @@ fn test_init_token() {
         pin_value.len() as CK_ULONG,
         std::ptr::null_mut(),
     );
-    assert_eq!(ret, CKR_PIN_INCORRECT);
+    if testtokn.dbtype == "nssdb" {
```
Comment (simo5) [inline src/tests/init.rs:44]: I would skip the fn_init_otken entirely if no nssdb, rather than matching just he result ?

File: src/tests/login.rs [hunk 3]
```diff
@@ -82,7 +82,12 @@ fn test_login() {
     let mut token_info = CK_TOKEN_INFO::default();
     let ret = fn_get_token_info(testtokn.get_slot(), &mut token_info);
     assert_eq!(ret, CKR_OK);
-    assert_eq!(token_info.flags & pin_flags_mask, CKF_USER_PIN_COUNT_LOW);
+    if testtokn.dbtype == "nssdb" {
```
Comment (simo5) [inline src/tests/login.rs:85]: Here too, I would just skip the operation, rather than matching to a fixed value.

File: src/tests/nssdb.rs [hunk 4]
```diff
@@ -175,6 +175,9 @@ fn test_nssdb_token() {
     let ret = fn_find_objects_final(session);
     assert_eq!(ret, CKR_OK);
 
+    #[cfg(not(feature = "rsa"))]
```
Comment (simo5) [inline src/tests/nssdb.rs:178]: Given nssdb always requires HMAC, what about changing some of these operations to use/generate a SECRET value usable by HMAC and defer RSA testing to the general RSA testing?

RSA is not really key here, what is ey is testing that encryption.decryption of a secret attribute works and fetching works as expected.

**Review by simo5 — APPROVED** [review]
Looks largely ok, only minor nits.
Thanks this is exactly where I wanted to see things headed!

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -75,39 +76,41 @@ jobs:
...
-          fi
-          if [ "${{ matrix.name }}" = "dynamic" ]; then
-            cargo build -vv --release --features standard,dynamic
+          FEATURES=""
+          OPTS=""
+
+          if [ "${{ matrix.db }}" = "memorydb" ]; then
+            FEATURES="memorydb"
+          elif [ "${{ matrix.db }}" = "jsondb" ]; then
+            FEATURES="memorydb"
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply .github/workflows/build.yml:87]: Indeed! Fixing.

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -75,39 +76,41 @@ jobs:
...
-          fi
-          if [ "${{ matrix.name }}" = "release" ]; then
-            cargo build -vv --release --features standard
-          fi
-          if [ "${{ matrix.name }}" = "dynamic" ]; then
-            cargo build -vv --release --features standard,dynamic
+          FEATURES=""
+          OPTS=""
+
+          if [ "${{ matrix.db }}" = "memorydb" ]; then
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply .github/workflows/build.yml:84]: Indeed. Did not think about that initially. Changing

File: src/tests/mod.rs [hunk 7]
```diff
@@ -129,8 +140,8 @@ impl TestToken<'_> {
         file.write(data.as_bytes()).unwrap();
     }
 
-    fn make_init_string(&self) -> String {
-        let confname = format!("{}.conf", self.filename);
+    fn make_init_string(&mut self) -> String {
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/tests/mod.rs:143]: Sounds like this is no longer needed and is some leftover from some initial attempts.

File: src/tests/mod.rs [hunk 8]
```diff
@@ -167,13 +178,40 @@ impl TestToken<'_> {
...
 
+    #[allow(unreachable_code)]
+    fn get_default_db(name: &str) -> (&'static str, String) {
+        use std::fmt::Write;
+
+        /* in order of preference, the first wins */
+        #[cfg(feature = "nssdb")]
+        {
+            let ret = format!("configDir={}/{}/", TESTDIR, name);
+            return ("nssdb", ret);
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/tests/mod.rs:189]: Good suggestion. Changed.

File: src/tests/objects.rs [hunk 9]
```diff
@@ -115,17 +115,19 @@ fn test_create_objects() {
...
-        CKO_CERTIFICATE,
-        &[(CKA_CERTIFICATE_TYPE, CKC_X_509)],
-        &[
-            (CKA_CHECK_VALUE, "ignored".as_bytes()),
-            (CKA_SUBJECT, "subject".as_bytes()),
-            (CKA_VALUE, "value".as_bytes())
-        ],
-        &[(CKA_TOKEN, true), (CKA_TRUSTED, false)],
-    ));
+    if testtokn.dbtype != "nssdb" {
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/tests/objects.rs:123]: Probably misread. Sorry. It should have been on the DATA import if I see right.

File: src/tests/aes.rs (hunk 1 above)
Comment (Jakuje) [reply src/tests/aes.rs:520]: I did not like either approach. With the `get_test_data`, the hex blocks get very indented, but huge blobs inside of function make the reading harder. In case of AES, the values are quite short so I considered this ok.

File: src/tests/init.rs (hunk 2 above)
Comment (Jakuje) [reply src/tests/init.rs:44]: Fine by me. I though keeping the test around and checking return value of the nssdb will make sure we do not regress even in these corner cases, but skipping is probably ok. Changing.

**Comment by Jakuje** [comment]
> The idea is to be comprehensive in the coverage, or just have at least some test coverage?
> 
> because I has some experience with [all-pairs testing](https://en.wikipedia.org/wiki/All-pairs_testing) and extension of it to triplets, quadruplets, etc....

Right now, we have around 25 features:

https://github.com/latchset/kryoptic/blob/main/Cargo.toml#L49

They have some internal dependencies and then there is the "default" feature which we want or dont want to build. I was worried that doing all pairs (or triplets) would explode or would be even more work to exclude meaningless ones.

**Comment by tomato42** [comment]
> > The idea is to be comprehensive in the coverage, or just have at least some test coverage?
> > because I has some experience with [all-pairs testing](https://en.wikipedia.org/wiki/All-pairs_testing) and extension of it to triplets, quadruplets, etc....
> 
> Right now, we have around 25 features:
> 
> https://github.com/latchset/kryoptic/blob/main/Cargo.toml#L49
> 
> They have some internal dependencies and then there is the "default" feature which we want or dont want to build. I was worried that doing all pairs (or triplets) would explode or would be even more work to exclude meaningless ones.

all pairs (n-tuples) testing is especially effective if you have a lot of features, as then you avoid the combinatorial explosion

and no, having meaningless combinations is not a problem, when you generate the combinations you can exclude combinations that are either impossible to support (like testing ECDH on build without ECC) or combinations we explicitly don't want to support (for whatever reason)

File: src/tests/objects.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/tests/objects.rs:113]: Ok, this was just in combination with FIPS mode for some reason
```
thread 'tests::objects::test_create_objects' panicked at src/tests/objects.rs:123:13:
```
I will have to dig into this deeper to figure out why it fails only in FIPS + NSSDB

**Comment by simo5** [comment]
> > > The idea is to be comprehensive in the coverage, or just have at least some test coverage?
> > > because I has some experience with [all-pairs testing](https://en.wikipedia.org/wiki/All-pairs_testing) and extension of it to triplets, quadruplets, etc....
> > 
> > 
> > Right now, we have around 25 features:
> > https://github.com/latchset/kryoptic/blob/main/Cargo.toml#L49
> > They have some internal dependencies and then there is the "default" feature which we want or dont want to build. I was worried that doing all pairs (or triplets) would explode or would be even more work to exclude meaningless ones.
> 
> all pairs (n-tuples) testing is especially effective if you have a lot of features, as then you avoid the combinatorial explosion
> 
> and no, having meaningless combinations is not a problem, when you generate the combinations you can exclude combinations that are either impossible to support (like testing ECDH on build without ECC) or combinations we explicitly don't want to support (for whatever reason)

In most cases each algorithm is fully independent from each other algorithm, so that testing them in pairs makes little sense.

We have a core that is always involved and therefore is always tested.

Then we have the algorithms which just operate on specific data, regardless of the core or anything else.

The only major variability is the databases, because what they contain may influence the result of specific tests.

So perhaps testing each "true leaf" algorithm on their own with each database may make some sense. Mostly to ensure there is no unexpected dependency between algorithms, but that is rather simple to verify so I would not want to spend a lot of CPU on that, and definitely not make CI much slower due to this.

Perhaps we can have matrix testing done once a week on a scheduled CI run just for completeness.

File: src/tests/objects.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/tests/objects.rs:113]: Ok, I tried to debug this with gdb and I have no idea why the combination of NSSDB and FIPS makes this panic (unless some weird optimization is going on there).

If I see right, we never reach the `fn_create_object()`, but the `import_object()` test helper does not do anything suspicious related to NSSDB nor FIPS.

File: src/tests/nssdb.rs (hunk 4 above)
Comment (Jakuje) [reply src/tests/nssdb.rs:178]: Indeed, I can do that so we can move forward with this change.

**Comment by Jakuje** [comment]
> warning: method `dbsuffix` is never used

With recent changes the dbsuffix is not used anywhere. I kept it for now because it might be useful in the future, even though it works only for SQL and JSON. If you do not agree, I can remove it altogether.

File: src/tests/objects.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (simo5) [reply src/tests/objects.rs:113]: Do you get back a specific error number?
I wonder if the pbkdf2 used to generate the key to encrypt the data fails because the salt is to short or something like that.

File: src/tests/nssdb.rs (hunk 4 above)
Comment (simo5) [reply src/tests/nssdb.rs:178]: Up to you.

**Comment by simo5** [comment]
> > warning: method `dbsuffix` is never used
> 
> With recent changes the dbsuffix is not used anywhere. I kept it for now because it might be useful in the future, even though it works only for SQL and JSON. If you do not agree, I can remove it altogether.

We can always resurrect from git, I am ok dropping anything that is not used anymore.

File: src/tests/objects.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/tests/objects.rs:113]: Not sure, if the 32 is the specific error:
```
thread 'tests::objects::test_create_objects' panicked at src/tests/objects.rs:123:13:
32
```
https://github.com/latchset/kryoptic/actions/runs/12236641232/job/34130555468?pr=122

That would map back to `CKR_DATA_INVALID`. Are there some parts of the certificates encrypted at all?

File: src/tests/nssdb.rs (hunk 4 above)
Comment (Jakuje) [reply src/tests/nssdb.rs:178]: I will drop it then. The current solution with if feature + return is ugly.

Comment (Jakuje) [reply src/tests/nssdb.rs:178]: Or try to move it to RSA test if possible.

File: src/tests/mod.rs [hunk 10]
```diff
@@ -167,13 +179,38 @@ impl TestToken<'_> {
         }
     }
 
+    #[allow(unreachable_code)]
+    fn get_default_db(name: &str) -> (&'static str, String) {
+        /* in order of preference, the first wins */
+        #[cfg(feature = "nssdb")]
+        {
+            let ret = format!("configDir={}/{}/", TESTDIR, name);
+            return (storage::nssdb::DBINFO.dbtype(), ret);
```
Comment (simo5) [inline src/tests/mod.rs:188]: Do we actually need to get this string by making modules public ?
At the end the dbtype() is nothing else than the db name which is what we are using to select which storage option we are going to fetch this same string from...
I'd like to keep the modules private to the storage tree, and accessory function would work as well if needed()

File: src/tests/objects.rs [hunk 11]
```diff
@@ -107,13 +107,15 @@ fn test_create_objects() {
...
     testtokn.login();
 
-    let _ = ret_or_panic!(import_object(
-        session,
-        CKO_DATA,
-        &[],
-        &byte_values,
-        &bool_values,
-    ));
+    if testtokn.dbtype != "nssdb" {
```
Comment (simo5) [inline src/tests/objects.rs:113]: This should not be necessary anymore

**Review by simo5 — CHANGES_REQUESTED** [review]
Only a couple of nits, but great job with this, thanks!!

File: src/tests/objects.rs (hunk 11 above)
Comment (Jakuje) [reply src/tests/objects.rs:113]: The CKO_DATA object type is not posible to store in the NSS DB if I understand it right.

File: src/tests/init.rs [hunk 12]
```diff
@@ -69,7 +69,7 @@ fn test_init_token() {
     );
 
     #[cfg(feature = "fips")]
-    {
+    if testtokn.dbtype != "nssdb" {
```
[the remark this answers is not in the document]
Comment (Jakuje) [reply src/tests/init.rs:75]: Yes, it is. This attempts to create the CKO_VALIDATION objects, that are not supported by the NSS DB either. Not sure if this is feature or bug though.

[the remark this answers is not in the document]
Comment (simo5) [reply src/tests/init.rs:75]: Oh right, currently they are synthetic objects in softoken, can't store them.

File: src/tests/mod.rs (hunk 10 above)
Comment (simo5) [reply src/tests/mod.rs:188]: Yeah by accessor I meant a function exposed in mod.rs ... but no biggie, it is ok as is for now

File: src/tests/objects.rs (hunk 11 above)
Comment (simo5) [reply src/tests/objects.rs:113]: Doh! Of course.. I spoke with Bob and we could actually store them, but to be able to that the nssdb.rs code will need to be changed, so nothing to do in this PR.

**Review by simo5 — APPROVED** [review]

**Comment by simo5** [comment]
Thanks for this work, really appreciated!
